pkg/client/injection/reconciler/samples/v1alpha1/addressableservice/reconciler.go

```diff
@@ -179,14 +190,107 @@ func (r *reconcilerImpl) updateStatus(existing *v1alpha1.AddressableService, des
 	})
 }
 
-// TODO: move this to knative.dev/pkg/reconciler
-// RetryUpdateConflicts retries the inner function if it returns conflict errors.
-// This can be used to retry status updates without constantly reenqueuing keys.
-func RetryUpdateConflicts(updater func(int) error) error {
-	attempts := 0
-	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		err := updater(attempts)
-		attempts++
+// updateFinalizersFiltered will update the Finalizers of the resource.
+// TODO: this method could be generic and sync all finalizers. For now it only
+// updates defaultFinalizerName.
+func (r *reconcilerImpl) updateFinalizersFiltered(ctx context.Context, resource *v1alpha1.AddressableService) error {
+	finalizerName := defaultFinalizerName
+
+	actual, err := r.Lister.AddressableServices(resource.Namespace).Get(resource.Name)
+	if err != nil {
 		return err
-	})
+	}
+
+	// Don't modify the informers copy.
+	existing := actual.DeepCopy()
+
+	var finalizers []string
+
+	// If there's nothing to update, just return.
+	existingFinalizers := sets.NewString(existing.Finalizers...)
+	desiredFinalizers := sets.NewString(resource.Finalizers...)
+
+	if desiredFinalizers.Has(finalizerName) {
+		if existingFinalizers.Has(finalizerName) {
+			// Nothing to do.
+			return nil
+		}
+		// Add the finalizer.
+		finalizers = append(existing.Finalizers, finalizerName)
+	} else {
+		if !existingFinalizers.Has(finalizerName) {
+			// Nothing to do.
+			return nil
+		}
+		// Remove the finalizer.
+		existingFinalizers.Delete(finalizerName)
+		finalizers = existingFinalizers.List()
+	}
```

Comment:
I'm wondering whether all of this should follow a similar pattern to the update status retry loop to avoid the initial get when it hasn't changed from under the cache.